shared/lib_battery_lifetime_lmolto.cpp

```diff
@@ -0,0 +1,183 @@
+        lifetime_t(rhs){
+    operator=(rhs);
+}
+
+lifetime_lmolto_t& lifetime_lmolto_t::operator=(const lifetime_lmolto_t& rhs) {
+    if (this != &rhs) {
+        *params = *rhs.params;
+        *state = *rhs.state;
+    }
+    return *this;
+}
+
+lifetime_t * lifetime_lmolto_t::clone() {
+    return new lifetime_lmolto_t(*this);
+}
+
+double lifetime_lmolto_t::runQcal() {
+    double SOC_avg = cycle_model->predictAvgSOC(state->cycle->DOD_max - state->cycle->DOD_min);
+    double q1 = q1_b0 * exp(q1_b1 * (1 / pow(state->lmo_lto->temp_avg, 3.)) * sqrt(SOC_avg)) *
+                      exp(q1_b2 * (1 / pow(state->lmo_lto->temp_avg, 2.)) * sqrt(SOC_avg));
+
+    double dqLossCal = 0;
+    if (state->lmo_lto->dq_relative_cal == 0.) {
+        if (state->day_age_of_battery > 0) {
+            dqLossCal = q1;
+        }
+    }
+    else {
+        dqLossCal = q1 * q2 * (double)pow(state->lmo_lto->dq_relative_cal * 0.01 / q1, (q2 - 1.) / q2);
+    }
+    state->lmo_lto->dq_relative_cal += dqLossCal * 100.;
+    return state->lmo_lto->dq_relative_cal;
+}
+
+double lifetime_lmolto_t::runQcyc() {
+    // use max DOD range regardless of if cycle has completed or not since using EFCs
+    double DOD_range = state->cycle->DOD_max - state->cycle->DOD_min;
+    double q3 = q3_b0 + q3_b1 * (double)pow(state->lmo_lto->temp_avg, 4.) * (double)pow(DOD_range, 0.25);
+
+    double dqLossCyc = 0;
+    if (state->lmo_lto->dq_relative_cyc == 0.) {
+        if (state->lmo_lto->EFC_dt > 0) {
+            dqLossCyc = q3 / state->lmo_lto->EFC_dt;
+        }
+    }
+    else {
+        if (state->lmo_lto->EFC_dt > 0) {
+            dqLossCyc = q3 * q4 * (double)pow(state->lmo_lto->dq_relative_cyc * 0.01 / q3, (q4 - 1) / q4);
+        }
+    }
+    // this NaN can happen if q3 <0 since pow with a negative base is a complex not real number.
+    // q3 < 0 when DOD_range ~0
+    if (std::isnan(dqLossCyc * state->lmo_lto->EFC_dt * 100))
+        dqLossCyc = 0.;
+    state->lmo_lto->dq_relative_cyc += dqLossCyc * state->lmo_lto->EFC_dt * 100;
+    return state->lmo_lto->dq_relative_cyc;
+}
+
+void lifetime_lmolto_t::integrateDegLoss() {
+    runQcal();
+    runQcyc();
+    state->q_relative = 100. - state->lmo_lto->dq_relative_cyc - state->lmo_lto->dq_relative_cal;
+
+    // reset daily cycle tracking
+    state->lmo_lto->EFC_dt = 0;
+    state->lmo_lto->temp_avg = 0;
+    cycle_model->resetDailyCycles();
+}
+
+void lifetime_lmolto_t::integrateDegParams(double dt_day, double prev_DOD, double DOD, double T_battery) {
+    double energy_throughput = fabs(DOD - prev_DOD) * 0.01 * 0.5;
+    state->lmo_lto->EFC_dt += energy_throughput;
+    state->lmo_lto->EFC += energy_throughput;
+    state->lmo_lto->temp_avg += T_battery * dt_day;
+    state->cycle->cum_dt += dt_day;
+}
+
+void lifetime_lmolto_t::runLifetimeModels(size_t lifetimeIndex, bool charge_changed, double prev_DOD, double DOD,
+                                          double T_battery) {
+    T_battery += 273.15;
```

Comment:
Unless I am mistaken T_battery is converted from Celsius to Kelvin for all the lifetime models. Are there other reasons that this input is being passed in Celsius (`thermal->T_battery()`) ?

Reply:
You're correct about that, but I think the reason Celsius is used and passed around rather than Kelvin is because of the thermal model, and the "cap_vs_temp" input, which is the relationship between the instantaneous thermal capacity and the temp in Celsius.